### PR TITLE
add `useHydration` check for ShadowRoot

### PR DESCRIPTION
**Pull request by mayank99**
## Changes

This PR adds a new `useHydration` hook to help determine hydration state, which is then used in `VisuallyHidden` and `ShadowRoot` to prevent unnecessary `null` renders on client. *(Follow-up to #1936, #1954, #1956, for #1930)*

To make the hydration state available everywhere without re-running the same code over and over, `HydrationProvider` is used. `useHydration` will reuse the context value if the code has already executed. This optimization is also useful when the server render only happens for a small part of the application (e.g. initial page / loading state), and then every subsequent navigation happens entirely on client.

The implementation itself relies on `useSyncExternalStore` to determine `"hydrating"` state (see [`getServerSnapshot` docs](https://react.dev/reference/react/useSyncExternalStore#parameters)) and `useEffect` to determine `"hydrated"` state (see [`useEffect` docs](https://react.dev/reference/react/useEffect#displaying-different-content-on-the-server-and-the-client)).

## Testing

Tested `ProgressRadial` and `VisuallyHidden` in playgrounds:
- no flash during pure client render (vite)
- no errors during server render + hydration (next, astro)

Also tested that React 17 is functional (see [commit](https://github.com/iTwin/iTwinUI/pull/1962/commits/d0f3aa10d6ebaca34151fc59946b96750d00c65f)). I did get a hydration warning in Nextjs, but no serious errors.

Made sure all existing tests pass.

## Docs

N/A

Added changeset to briefly mention the overall effect on users.

### CODE REVIEW

File: packages/itwinui-react/src/core/utils/providers/HydrationProvider.tsx [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    <HydrationContext.Provider value={isHydrated}>
+      {!isHydrated ? <HydrationCheck onHydrate={onHydrate} /> : null}
+      {children}
+    </HydrationContext.Provider>
+  );
+};
+
+/** This is extracted into a child component to ensure it runs first. */
+const HydrationCheck = ({ onHydrate }: { onHydrate: () => void }) => {
+  React.useEffect(() => void onHydrate(), [onHydrate]);
```
Comment (mayank99) [inline packages/itwinui-react/src/core/utils/providers/HydrationProvider.tsx:64]: This is a react thing. Child effects always run first
image

File: packages/itwinui-react/src/core/utils/components/ShadowRoot.tsx [hunk 2]
```diff
@@ -85,27 +86,34 @@ function useShadowRoot(
...
-      // Flush the state immediately to ensure layout measurements in parent component are correct
+      // Flush the state immediately after shadow-root is attached, to ensure that layout
+      // measurements in parent component are correct.
+      // Without this, the parent component may end up measuring the layout when the shadow-root
+      // is attached in the DOM but React hasn't rendered any slots or content into it yet.
       ReactDOM.flushSync(() => setShadowRoot(shadow));
+    };
+
+    queueMicrotask(() => {
+      setupShadowRoot();
```
Comment (mayank99) [inline packages/itwinui-react/src/core/utils/components/ShadowRoot.tsx:116]: By fixing the first client render, I was able to reproduce the original issue related to incorrect measurements in Dialog. It was trying to read measurements when the shadow-root was attached but state wasn't updated.

image

To fix it, I moved both of these things (`attachShadow` browser call and `setShadowRoot` react call) closer to each other into queueMicrotask. (Same as it was before #1954)

File: packages/itwinui-react/src/core/utils/providers/HydrationProvider.tsx [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+ * @private
+ */
+export const useHydration = () => {
+  // Returns true only if getServerSnapshot is called on the client.
+  // In React 18, this is true during hydration.
+  const hydrating = useSyncExternalStore(
+    noopSubscribe,
+    () => false,
+    () => !isServer,
+  );
```
Comment (rohan-kadkol) [inline packages/itwinui-react/src/core/utils/providers/HydrationProvider.tsx:29]: I'm a little confused about why we're using `useSyncExternalStore` here.

Comment (mayank99) [reply packages/itwinui-react/src/core/utils/providers/HydrationProvider.tsx:29]: `useSyncExternalStore` is the only way to check for hydration in react.

- `noopSubscribe` is just an empty function. It needs to be referentially stable, so it's declared outside of the component.
- `getSnapshot` returns false (called during non-hydration client renders)
- `getServerSnapshot` returns true on the server and during hydration. So I'm using a boolean check to make it only return true during hydration.

**Pull request merged**